### ISSUE TITLE
Implement an abstraction for wamp transports

### ISSUE THREAD

**Issue by davidchappelle** [crossbario/autobahn-cpp#54]
The session API is templated and requires that both an input stream and output stream be provided. For example:

    autobahn::session<boost::asio::local::stream_protocol::socket, boost::asio::local::stream_protocol::socket>
    autobahn::session<boost::asio::ip::tcp::socket, boost::asio::ip::tcp::socket>
    autobahn::session<int, int>

Calls to *boost::asio::write*, *boost::asio::read*, and *boost::asio::async_read* are then issued internally as required to send and receive messages. This is restricting such that the templated types must be compatible with *boost::asio* semantics for reading and writing data. In the near future, support will likely be added for Websocket, HTTP/2, and a native application transport to call directly into a WAMP router. In order to do this, we will need to be able to have the session API internally write to these types of transports. As a result we need to provide further abstraction in place of the templated streams that are provided to the session. For example:

    class wamp_stream
    {
    public:
        wamp_stream();
        virtual wamp_stream();

        virtual int read(char* buffer, int length) = 0;
        virtual void async_read(char* buffer, int length, read_handler handler) = 0;
        virtual void write(const char* buffer, int length) = 0;
    };
 
    class wamp_websocket_stream : public wamp_stream
    {
    ...
    };

    class wamp_http2_stream : public wamp_stream
    {
    ...
    };

    class wamp_native_stream : public wamp_stream
    {
    ...
    };

Then we can change the session API to no longer require templates for the input and output streams and instead pass in objects of the desired types:

    autobahn::session(boost::asio::io_service& io, wamp_stream& in, wamp_stream& out, bool debug=false)

As an added benefit, this would also allow the session API to participate in a factory pattern which could be desireable for implementations requiring dynamic session generation based on transport type.

**Comment by davidchappelle** [comment]
I am going to start work on this as I now need this to avoid a bunch of template issues.

**Comment by oberstet** [comment]
> This is restricting such that the templated types must be compatible with boost::asio semantics for reading and writing data.

So the semantics of above `wamp_stream` abstraction would be different from a asio stream? In what way? Because `wamp_stream` basically is a byte stream abstraction, right?

**Comment by davidchappelle** [comment]
@oberstet by abstracting an interface we can actually hide implementation details for things like native c++-to-c++ router calls, websocket communication, http2 communication, etc. Also, having template parameters on the session class is a bit limiting. For example, if you look at how I did components, I had to templatize the component class. As a result, I can't have something generically accept a component. For example:

    class system_client
    {
    public:
        my_client(autobahn::wamp_tcp_component& component);

        status get_network_interfaces(network_interfaces& interfaces);
        .
        .
        .
    };

If I want to reuse this client API I would need to templatize it to accept the component type and even then it limits me to static types. Alternatively, if we could get rid of the template requirement I could write this simply as:

    class system_client
    {
    public:
        my_client(autobahn::wamp_component& component);

        status get_network_interfaces(network_interfaces& interfaces);
        .
        .
        .
    };

Taking this approach I can then pass whatever component type I want to the client without the need for templates. This allows me to reuse the client in a number of scenarios where I may be using TCP, UDS, Websocket, Http2, etc as the transport type and I can determine this dynamically rather than statically.

**Comment by davidchappelle** [comment]
From reading through *boost/asio/read.hpp* and *boost/asio/write.hpp* once could implement an object with an interface that supported the expected API of read_some, async_read_some, write_some, async_write_some. However, it still leaves us stuck with templates which force the templatization of types up to the topmost user apis as illustrated in my example above.

**Comment by davidchappelle** [comment]
@oberstet any other insights here? I am leaning towards just leaving things templated for now until we have a more compelling need.

**Comment by oberstet** [comment]
@davidchappelle I think the right transport abstraction would probably more send/receive whole messages. So essentially, what the spec talks about with that bidirectional, reliable message channel. A standard ASIO stream pair would be wrapped in a class we provide, which implements a transport interface that is message based. And then this class is given as a template parameter. Maybe something like this works?

**Comment by oberstet** [comment]
I am also wondering if the transport should be a template parameter anyway, or not more sort of a list of transport things that implement the transport interface. Alllowing run-time polymorphism at least here would be neat.

**Comment by davidchappelle** [comment]
Just getting back to this. I'll draft a commit based on the discussion above.

**Comment by davidchappelle** [comment]
I think I understand your point better now about abstracting in terms of a message based transport. I should have something ready later today that is much more in line with what you are describing.

**Comment by oberstet** [comment]
@davidchappelle awesome!

FWIW, eg in AutobahnPython, WAMP transports are abstracted via 2 interfaces:

* `ITransport` (https://github.com/tavendo/AutobahnPython/blob/master/autobahn/wamp/interfaces.py#L123)
* `ITransportHandler` (https://github.com/tavendo/AutobahnPython/blob/master/autobahn/wamp/interfaces.py#L171)

Serializers are created (or given to) concrete transports, but the abstract transport interface just shuffles WAMP messages in their "parsed" form (eg https://github.com/tavendo/AutobahnPython/blob/master/autobahn/wamp/message.py#L222)

In AutobahnJS, we also have a message-based transport abstraction, eg https://github.com/tavendo/AutobahnJS/blob/master/package/lib/transport/websocket.js#L45 and the sibling files in that folder.

**Comment by davidchappelle** [comment]
Does it still make sense for session to take in an in/out transport? Seems that by abstracting a transport it can be up to the transport itself how to manage its in/out streams.

**Comment by oberstet** [comment]
No, in hindsight, I think this is over-engineering. Also, the lifetimes of Sessions and Transports should be completely decoupled (the Session ctor does NOT take a Transport and vice versa) - another lesson learned. Let me think for some minutes .. will post here again.

**Comment by oberstet** [comment]
I'd approach it like this:

1. define 2 pure abstract base classes `ITransport` and `ITransportHandler`
2. `ITransport` has `virtual void send(const wamp_message&) = 0`, `virtual bool is_open() const = 0` and `virtual void attach(ITransportHandler&) = 0`
3. `ITransportHandler` has `virtual void on_attach(ITransport&) = 0`, `virtual void on_message(const wamp_message&) = 0` and `virtual void on_detach(bool was_clean, std::string close_reason) = 0`

Then, `Session` implements `ITransportHandler`, and concrete Transports (eg WAMP-over-WebSocket using MsgPack) implement `ITransport`.

When the boilerplate has created and opened a transport, it attaches the transport to a session (which is created completely independent of the transport). The session in on_attach will only save a reference/pointer to the transport. Then, when calling session.join, the session will use ITransport to send messages, and get on_message called and process incoming messages.

When the transport looses it's connection, it will _also_ call on_detach. But a session may also be detached from a transport without the transport being closed. etc etc

Before jumping into implementation, we should hash out above. The implementation is more or less straightforward. The hard part is getting above right.

**Comment by oberstet** [comment]
More braindumping: in AB C++, WAMP messages are abstracted here https://github.com/tavendo/AutobahnCpp/blob/master/autobahn/wamp_message.hpp

This isn't really an abstraction of WAMP messages: it's just a array of msgpack objects. Not every array of msgpack objects is a valid WAMP message.

That is, right now, transport would be responsible for 1) framing streams of bytes into message frames and 2) parsing those according to serializer in use into msgpack object arrays.

But the further parsing of arrays of msgpack objects into real WAMP messages happens inside Session, and there isn't even classes or what for say `HELLO` or `CALL` (the WAMP messages).

With AB Py, transports are responsible for 1), 2) _and_ 3)!

This is important for the following reason: when running an AB Py WAMP component "side-by-side" (inside CB), the "transport" used between the component and the router boils down to simple function calls. When 3) isn't done in the transport, the session will still need to run "parse" and check code before actually performing the action.

Note that I am not saying one or the other is strictly better, I just want to raise awareness to these aspects.

Unrelated sidenote: what is this zone stuff eg here https://github.com/tavendo/AutobahnCpp/blob/master/autobahn/wamp_session.ipp#L805?

**Comment by davidchappelle** [comment]
I see where you are going with this @oberstet. This would allow us to cycle through a list of transports trying to establish a session as well as recycle a transport for use by multiple successive sessions. Should be pretty straight forward.

As for the zone stuff, msgpack uses a concept called a zone. When you create message pack objects that do not fit into the intrinsic type category, the object requires a zone in order to allocate memory. There is no way to deallocate using a zone though. So a zone cannot have an indefinite scope as it would just continue to grow unbounded. As a result, zones should only really live as long as the objects that were allocated from it.

**Comment by oberstet** [comment]
> I see where you are going ..

Exactly. These are all lessons (I think) I've learned;)

Further, the WAMP spec has a section about relation between WAMP session and transports. But this is unnec. limiting. E.g. when we get session freeze/resume, a session can certainly outlive a transport. So I'd vote for removing this from the spec altogether (btw: if you can shelve out at least a little time, pls respond on the Hybi list. please.) Reusing a transport with different sessions was already possible anyway (and AB Py exposes it, but not AB JS).

**Comment by oberstet** [comment]
another note rgd `ITransportHandler`: I have named the callbacks `on_attach` and `on_detach` - NOT `on_open` and `on_close`. The latter would suggest the transport connection would have opened or closed. But that isn't the point. The transport might have been opened long ago and used for whatever, but then is _attached_ to a transport consumer (`ITransportHandler`). Of course a transport can only be attached to 1 transport handler at a time .. but.

**Comment by davidchappelle** [comment]
Hbyi?

**Comment by oberstet** [comment]
@davidchappelle https://mailarchive.ietf.org/arch/search/?email_list=hybi .. the threads on WAMP. I have posted a draft RFC (https://tools.ietf.org/html/draft-oberstet-hybi-tavendo-wamp-00) - but unless we can show significant interest by multiple parties in developing this RFC, the Hybi/IETF chairs will likely reject it soon! I perfectly understand that you have tight time budget (me too), but this is a critical phase with WAMP. And there is a window of opportunity - which will close again (because the Hybi/IETF chairs want to close the WG).

**Comment by jpetso** [comment]
Good discussion here, I like the direction that this is (hopefully) going.

I was just going to say that I don't see a strong argument to introduce two interfaces where one might be sufficient - what you call `ITransport` is the thing that needs to be abstract and will be the base of various implementations, but `ITransportHandler` is apparently just the session and nothing else. Why do we want to make an interface for the session when (a) `ITransport` targets WAMP specifically and (b) the session is the only attachment target? 

Also, maybe just to clarify this point, I'm under the assumption that `ITransport` & Co. are just working titles since they don't match too well with the current naming style in AutobahnCpp (which is more like STL/Boost and uses neither camel-casing nor `I*` prefixes). Correct?

**Comment by davidchappelle** [comment]
Yes I am sticking to current naming conventions. The abstraction of
`wamp_transport_handler` is helpful to break the cyclic reference between
wamp_session and wamp_transport.

**Comment by davidchappelle** [comment]
FWIW, this should make adding support for native transports much simpler
and improve performance since full serialization/copy will not be required.

**Comment by davidchappelle** [comment]
I have some questions around implementation details for detaching transport handlers. It seems that ITransport also needs to have a method to detach the ITransportHandler:

    void ITransport::detach(bool was_clean, const std::string& reason)

What is the transport handlers implementation supposed to do when `on_detach` is called? If `was_clean` is `true` should it try and leave the session? What about if `was_clean` is `false`?

**Comment by davidchappelle** [comment]
I took a look at AB Py for some clues but it looks like is using the old API `on_open` and `on_close`.

**Comment by oberstet** [comment]
Could you create a 1st PR with only the 2 abstract bases for the interfaces and nothing else? We can describe the semantics of the interface in the docstrings of the respective methods.  And discuss the semantics of on_detach and such based on a concrete proposal. When we've done that, the implementation will fall into place I guess. Sounds good?

**Comment by davidchappelle** [comment]
Pull request submitted as PR #58.

On a side note, I can't find a way to post anything to that thread on Hybi.
Seems to require a username/password to login which I don't have and can't
figure out how to create one.

**Comment by goeddea** [comment]
To participate in the hypi mailing list, you need to be a subscriber. Subscription is quick and painless - https://www.ietf.org/mailman/listinfo/hybi

**Comment by oberstet** [comment]
@davidchappelle I merged PR #58 and now do some changes. Let's continue the discussion here ..

Rgd. the Hybi list: does it work for you now?

**Comment by oberstet** [comment]
Regarding `wamp_transport_handler.on_message` and `wamp_transport.send`: the one question we need to resolve is `wamp_message` (and hence is transparent to the interface definition formally, but still): will this be a list of msgpack objects (as it is right now), or will it be a proper, full parsed WAMP message abstraction?

**Comment by oberstet** [comment]
The issue 2) is about flow-control:

Consider a session that wants to issue calls to a procedure "as fast as possible".

A simple loop that will issue calls without waiting for the previous call to return might overwhelm the transport or the callee.

A loop that will issue calls one after the other (that is, only issue a new call when the last has returned) won't overwhelm the transport or callee, but might run slower than possible.

**Comment by davidchappelle** [comment]
@oberstet I have made wamp_message a simple class for the time being. It
simply wraps a vector of msgpack objects and a zone for the time being.
Eventually we could do what I did for bonefish and make each wamp message
type a fully parsed concrete type but what I have now should suffice for
the time being until we get the current abstraction work finished.

Flow control should be easy to manage by the user by simply waiting for
each future to resolve before issuing the next request. I think this
provides the most amount of flexibility.

I am in the process of updating all of the wamp message processing in the
session class to now work with the message class.

**Comment by oberstet** [comment]
@davidchappelle I agree the problem 1) can be postponed.

We should get the `wamp_transport.send` and `wamp_transport_handler.on_message` right first. This is a tricky one.

Currently, we use `boost::asio::write` under the hood. If a transport is using this to implement `wamp_transport.send`, a slow receiver which excerts TCP backpressure will simply make the program halt, since the socket write call is blocked in the kernel when the outgoing TCP buffers get full. IOW: `wamp_transport.send` might block, which is bad of course.

FWIW, the solution that Twisted implements is: a write _never_ blocks. The data is simply buffered in user space. Of course the question then is: how to prevent memory exhaustion. The Twisted answer to this is Twisted Producer/Consumers.

**Comment by oberstet** [comment]
> simply waiting for each future to resolve

Which future? `wamp_transport.send` returns void ..

**Comment by davidchappelle** [comment]
The user must eventually wait on futures returned by the session API for
things like calling/publishing/etc to complete. For most applications this
may be sufficient for achieving flow control.

Yes any tcp backpressure would be potentially problematic. In the current
implementation since we post all message to the io service in the session
object, the io service would basically act as a buffering mechanism. The
write itself is done in the context of the io_service. I think that
selecting a better mechanism for queuing writes may be able to be done as
an enhancement?

**Comment by oberstet** [comment]
Publish is unacknowledge by default, so there is no future at the session level to wait on.

FWIW, here is a sketch of how the Twisted way would roughly translate:

```c++
class wamp_transport
{
public:
    /*!
     * Send the given WAMP message over the transport to the peer.
     *
     * This method does _not_ block. An implementation of this method will
     * usually serialize the WAMP message to a byte string, frame the bytes
     * according to the WAMP transport framing in place (e.g. WebSocket or
     * RawSocket) and then buffer and arrange for asynchronous sending of
     * the bytes, but return immediately.
     *
     * Note that when you first send message 1 and then message 2, _and_
     * message 2 was indeed received by the peer, it is guaranteed that
     * a) message 1 was also received, and b) that the order of receiving
     * first message 1 and then 2 is preserved. Hence a transport guarantees
     * strict message ordering and a slightly weaker form of reliability.
     *
     * @param message The message to be sent.
     */
    virtual void send(const std::shared_ptr<wamp_message>& message) = 0;

    /*!
     * Pause receiving messages on the attached transport handler. This is
     * used by transport handlers to excert backpressure on the sending peer.
     */
    virtual void pause() = 0;

    /*!
     * Resume receiving messages on the attached transport handler. This is
     * used by transport handlers to signal readiness to consume more messages
     * and lift backpressure from the sending peer.
     */
    virtual void resume() = 0;
};

class wamp_transport_handler
{
public:
    /*!
     * Called by the transport when a message is received.
     *
     * @param message The message that has been received.
     */
    virtual void on_message(const wamp_message& message) = 0;

    /*!
     * Called by the transport when outgoing buffers are filling up (have reached
     * the high watermark of buffered data) and this transport handler should (temporarily)
     * stop sending messages.
     */
    virtual void on_pause() = 0;

    /*!
     * Called by the transport when outgoing buffers are getting empty (have reached
     * the low watermark of buffered data) and this transport handler may resume
     * sending messages.
     */
    virtual void on_resume() = 0;
};
```

**Comment by davidchappelle** [comment]
I like that API. what would be the behaviour of the session object when
`on_pause()` is called?

**Comment by oberstet** [comment]
When on_pause was called, the session goes into "paused" state. In "paused" state, any call to `session.call`, `session.publish`, `session.register` and `session.subscribe` will raise a `transport_paused` exception.

**Comment by oberstet** [comment]
Here is an alternative design:

```c++
class wamp_transport
{
public:
    /*!
     * Send the given WAMP message over the transport to the peer.
     *
     * The transport will resolve the returned future as soon as it is able
     * and willing to send more messages. This mechanism is used by the
     * transport to excert backpressure on the transport handler from the
     * receiving peer.
     *
     * @param message The message to be sent.
     */
    virtual boost::future<void> send(const std::shared_ptr<wamp_message>& message) = 0;
};

class wamp_transport_handler
{
public:
    /*!
     * Called by the transport when a message is received.
     *
     * The transport handler should resolve the returned future as soon as it is
     * able and willing to consume more messages. This mechanism can be used by
     * the transport handler to excert backpressure on the sending peer.
     *
     * @param message The message that has been received.
     */
    virtual boost::future<void> on_message(const wamp_message& message) = 0;
};
```

**Comment by davidchappelle** [comment]
The downside with this approach is that it introduces a lot of additional
locking overhead for each message sent.

**Comment by oberstet** [comment]
Locking? Why? This whole stuff should be single-threaded anyway?

---

Here is another (somewhat orthogonal) Q we need to decide:

a) have `wamp_transport` and `wamp_transport_handler` OR
b) only have `wamp_transport` and ..

let `wamp_transport` provide a listener pattern, so I can hook 1 or more listeners for: `on_message`, `on_pause` etc

**Comment by davidchappelle** [comment]
Futures are implemented with mutexes and condition variables. So using them
results in a bunch of locking overhead even if you are in a single threaded
context. Albeit an uncontended mutex will probably benefit from futex type
behaviour.

Not sure if multiple listeners would be very useful. If that type of
behaviour is desired it could be left up to the user of the session api to
`fan out` to multiple `listeners`.

**Comment by davidchappelle** [comment]
I think that we should we break this work up into multiple deliverables as
the code changes are starting to get to be fairly significant and we can
deliver it in smaller incremental changes.

1. Wamp transport abstraction

- Implement core abstraction for wamp transports to decouple rawsocket
transports from sessions.
- This would NOT include `on_attach` and `on_detach` behaviour just the
simple concept of transports.

2. Implement transport decoupling

- Implement `on_attach` and `on_detach` behaviour for
`wamp_transport_handler` as well as the new proposed API for initializing
transports/sessions (i.e. setup or main style).
- This will eliminate some boiler plate and allow for transport and session
reuse.

3. Implement flow control for transports

- Implement the `on_suspend` and `on_resume` for `wamp_transport` and
`wamp_transport_handler`.
- This would provide some flow control so that the session cannot be
overwhelmed with requests.

**Comment by oberstet** [comment]
I think I disagree on this one: IMO we should _first_ _define_ the proper interface.

The implementation can lack behind the interface, but the interface should be designed and fixed now. I'd prefer a half-baked implementation of a good and stable interface over an implementation of a half-baked interface.

> the code changes are starting to get to be fairly significant 

What are you implementing? The interface isn't "done", so there isn't anything to implement (or you are chasing a moving target) - maybe we have different modes of working;)

**Comment by davidchappelle** [comment]
There is a bunch of work to be done regardless of the interface(s) so I
have been tackling that between discussions. I have also created
`wamp_rawsocket_transport` and extracted all of the `wamp_session` related
code into that class. Basically all of the message framing/receiving etc.
Most of the interface that we have talked about at this point is just
stubbed out for the time being. Might be easier to get my refactoring in at
a minimum and then focus on implementing the API internals.

**Comment by oberstet** [comment]
> b) only have wamp_transport and ..

```c++
class wamp_transport
{
public:
    /*
     * SENDER interface
     */

    /*!
     * Send the given WAMP message over the transport to the peer.
     *
     * @param message The message to be sent.
     */
    virtual void send_message(const std::shared_ptr<wamp_message>& message) = 0;

    /*!
     * Listener type for "on_pause" events.
     */
    typedef std::function<void()> on_pause_t;

    /*!
     * Add a listener to "on_pause" events.
     */
    virtual void on_pause(on_pause_t& listener) = 0;

    /*!
     * Listener type for "on_resume" events.
     */
    typedef std::function<void()> on_resume_t;

    /*!
     * Add a listener to "on_resume" events.
     */
    virtual void on_resume(on_resume_t& listener) = 0;


    /*
     * RECEIVER interface
     */

    /*!
     * Listener type for "on_message" events.
     */
    typedef std::function<void(const wamp_message&)> on_message_t;

    /*!
     * Add a listener to "on_message" events.
     */
    virtual void on_message(on_message_t& listener) = 0;

    /*!
     * Pause receiving messages and firing "on_message" events. This is
     * used to excert backpressure on the sending peer.
     */
    virtual void pause() = 0;

    /*!
     * Resume receiving messages and firing "on_message" events. This is
     * used to signal readiness to consume more messages and lift backpressure
     * from the sending peer.
     */
    virtual void resume() = 0;
};
```

**Comment by davidchappelle** [comment]
So the `on_message` handler here is replacing
`transport_handler::on_message`?

**Comment by oberstet** [comment]
Yeah. This proposal is to further illustrate the design space: doing away with transport_handler altogether.

I am still exploring possibilities.

Above last proposal doesn't work quite that way: C++ makes it hard to have pointers to methods on an object (the implicit "this").

e.g. `virtual void on_message(on_message_t& listener) = 0;` does not work with a possible `wamp_session::on_message` implementation (I admit, I am spoiled by Python).

**Comment by oberstet** [comment]
fwiw, here is the last proposal with slightly better/readable naming:

* https://github.com/tavendo/AutobahnCpp/blob/transport_ifc/autobahn/wamp_transport.hpp#L41

And here is my current sketch of retaining 2 interfaces:

* https://github.com/tavendo/AutobahnCpp/blob/transport_ifc/autobahn/wamp_transport2.hpp
* https://github.com/tavendo/AutobahnCpp/blob/transport_ifc/autobahn/wamp_transport_handler.hpp

**Comment by oberstet** [comment]
one last note (need some sleep): I have thought about _why_ would one need flow-control at the WAMP transport level (when with most WAMP actions, there is feedback at the WAMP level .. call result etc on which one could hook on flow control).

Here is my list:

1. unacknowledged publications
2. progressive results
3. progressive calls (not yet there, but we'll have it)

Plus: 4. throttling calls based on call result won't run at maximum speed (since the roundtrip latency isn't hidden and used for actual work).

Of these, I think 2 and 3 are of real practical use (because the whole point of the "progressive" stuff is about sending/receiving larger amounts of data at the WAMP level).

1. can be circumvented using acknowledged publications.

And 4. is probably a minor thing .. but still.

**Comment by davidchappelle** [comment]
Should be possible with std::bind. For example:

m_transport.set_message_handler(std::bind(&wamp_session::on_message, this,
std::placeholders::_1));

**Comment by davidchappelle** [comment]
4) is less than ideal as you point out, the pipe will not be full and
latency accumulation will be very limiting for large amounts of data.

I am wondering if flow control can be an optional feature. For simple
applications it is likely insignificant and having the odd write block may
not be of any real concern. Or for cases where latency is a non issue and
throughput is high like loopback, unix domain sockets, and native
communication.

**Comment by davidchappelle** [comment]
I am going to cleanup what I have so far and push to my fork. Then you can
see what I have done to cleanup messages and decoupling transports from
sessions. Once we have settled on an API it should be very simple to
incorporate those api changes.

**Comment by oberstet** [comment]
@davidchappelle Thanks for the PR! It's unfortunate that I can commit more time to this right now .. I hope I can get some high-prio blockers out of the way soon, and the dive into C++ again.

Sidenote: we've published a new Internet-Draft: https://tools.ietf.org/html/draft-oberstet-hybi-tavendo-wamp-02

This now contains both the basic and the advanced profile. Hashing out this RFC is one of the high-prios on my list. Some of the features in Crossbar.io have surpassed the spec, and I want to bring those back in line ..

**Comment by davidchappelle** [comment]
How would you like to proceed on the wamp transport front? The transport API is how we had discussed and amended above. I apologize for the large commit but once you swap out the transport a number of other changes need to be made. So there shouldn't be many surprises there. For example, the wamp message. We needed to change the message structure a bit which resulted in a bunch of corresponding changes in all of the message marshalling/unmarshalling. This is one of the reasons why the commit looks large. There are lots of incidental changes like this that needed to be made. FWIW, we are using what I currently have on my fork davidchappelle/feature/wamp_transport_abstraction quite extensively in a production environment.

**Comment by davidchappelle** [comment]
I am going to decompose my commit into several smaller commits that will be easier to review.
